### ISSUE TITLE
Fix illegal reflective access operations on Java 9+

### ISSUE THREAD

**Issue by Kaerol** [j-easy/easy-random#385]
**openjdk 11.0.5 2019-10-15**
OpenJDK Runtime Environment (build 11.0.5+10-post-Ubuntu-0ubuntu1.118.04)
OpenJDK 64-Bit Server VM (build 11.0.5+10-post-Ubuntu-0ubuntu1.118.04, mixed mode, sharing)

**gradleVersion = '5.6'**

_./gradlew clean build_

**ISSUE**
```
[...]
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jeasy.random.util.ReflectionUtils (file:/home/programista/.gradle/caches/modules-2/files-2.1/org.jeasy/easy-random-core/4.1.0/58699e714ee8d1b4be51de7a4db76bf5911a9b39/easy-random-core-4.1.0.jar) to field java.lang.Throwable.backtrace
WARNING: Please consider reporting this to the maintainers of org.jeasy.random.util.ReflectionUtils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[...]
```

**REASON**
All illegal access operations will be denied in a future release

**Comment by benas** [comment]
We use maven in Easy Random. When I build the project with OpenJDK 11 (`mvn clean install`), I don't see this warning. May be our test suite does not cover your use case? Can you please provide a minimal project (even with gradle) that reproduces the issue? Otherwise please share the type that your are trying to randomize and a corresponding test that shows this warning.

**Comment by benas** [comment]
I think this is related to #243 . Starting from Java 10, granting access to objects should be explicit using the `--permit-illegal-access` flag.

@Kaerol Could you please try that and see if it removes the warning? Otherwise please provide a minimal example that reproduce the issue I will take a look.

**Comment by Kaerol** [comment]
Hi
I added to 
`gradle.properties` 
```
org.gradle.jvmargs=--illegal-access=permit
```
and to 
`build.gradle` 
```
tasks.withType(Test) {
    jvmArgs '--illegal-access=permit'
}
```
Nothing happen/changes acording to this article 
https://www.logicbig.com/tutorials/core-java-tutorial/modules/illegal-access-operations.html
`permit` is default.

When changed to `deny` I got an error 
```
      org.jeasy.random.ObjectCreationException: Unable to create a random instance of type class ....
      Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private transient java.lang.Object java.lang.Throwable.backtrace accessible: module java.base does not "opens java.lang" to unnamed module 
```
So it looks like place to switcher is good.

**Comment by benas** [comment]
Thanks. Can you share a minimal project please? Or at least the type that you are trying to randomize and that creates this warning.

**Comment by Kaerol** [comment]
Sure,
you were invited to GitHub Repo :)

**Comment by benas** [comment]
@Kaerol Thanks for sharing the repo.

I add your minimal test here for reference (I just changed the expected value in the assertion):

```java
class SomeException extends RuntimeException {
    SomeException(final UUID id) {
        super("Exception for object with identifier '"+id+"'.");
    }
}

@Test
void should() {
    // given
    EasyRandom easyRandom = new EasyRandom();

    // when
    final SomeException someException = easyRandom.nextObject(SomeException.class);

    // then
    assertEquals("eOMtThyhVNLWUZNRcBaQKxI", someException.getMessage());
}
```

Running this test with Jdk 11 leads to this warning:

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jeasy.random.util.ReflectionUtils (file:/Users/mbenhassine/projects/benas/easy-random/easy-random-core/target/classes/) to field java.lang.Throwable.backtrace
WARNING: Please consider reporting this to the maintainers of org.jeasy.random.util.ReflectionUtils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

I tried adding `--permit-illegal-access` as a JVM option but this seems to be not recognized:

```
Unrecognized option: --permit-illegal-access
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

This is what I read from the [official docs](https://docs.oracle.com/javase/9/tools/java.htm#JSWOR624):

```
--illegal-access=parameter
Note:
This option is a new option in JDK 9 and may not be available in future JDK versions.
```

So may be the option is not available anymore. That said, adding the JVM option `--add-opens java.base/java.lang=ALL-UNNAMED` (as mentioned in #243  [here](https://github.com/j-easy/easy-random/issues/243#issuecomment-282871162) and on SO [here](https://stackoverflow.com/a/46671473/5019386)) fixes the warning.

So I will let you configure your gradle build accordinlgy (I'm not familiar with gradle/groovy syntax, but this might help: https://discuss.gradle.org/t/unrecognized-jvm-option-for-add-opens-gradle-6-and-jdk11/34325).

Bottom line, Easy Random uses reflection to access your types, and starting from Java 9, you need to explicitly open your packages/modules (`java.base` in this case since the randomized type extends `RuntimeException`).

**Comment by Kaerol** [comment]
Hmmm,

maybe I'm wrong but `WARNING: All illegal access operations will be denied in a future release`
so it means that someday it may stop working...

If we disable warn ok its clear but not solve problem...

About remove this warning here is nice article...
https://dev.to/erichelgeson/removing-illegal-reflective-access-warnings-in-grails-4-393o

Put only 
```
def jvmOpenModulesArgs = [
        "--add-opens=java.base/java.util=ALL-UNNAMED",
]
```
Don't solve problem. 
I don't checked which options are required to remove that warning - because I don't think it good idea.

**Comment by benas** [comment]
Easy Random (and similar libraries using reflection) will stop working _unless_ you explicitly open your classes for reflective access. Strong encapsulation is a key motivation behind the module system introduced in Java 9, which is the best way IMO to prevent libraries/frameworks do all kinds of hacks to user types and break encapsulation.

As a class designer, when I create a private field in my type, I want this field to be private and inaccessible outside the class. Prior to Java 9, any library (including Easy Random) can use reflection to change the visibility of my private field (`Field#setAccessible(true)`), change its value, etc.. In other words, reflection breaks encapsulation. True OO design starts with Java 9: You can't touch my type unless I open it for you.

So as mentioned previously, starting from Java 9, you need to explicitly open your packages/modules so that Easy Random can work properly.

That said, we need to update Easy Random to replace the usage of [Field.isAccessible](https://github.com/j-easy/easy-random/blob/62b58cb1daa4f4499e1082d7cae175748a02a70c/easy-random-core/src/main/java/org/jeasy/random/util/ReflectionUtils.java#L154-L155) (which is deprecated starting from Java 9) with `field.canAccess(object)`. So let me re-open this issue for further consideration. Note that this will require an upgrade of the minimum java version to 9 (to be able to use `field.canAccess(object)`), so this will wait until the next major version v5 (And I don't exclude making v5 require Java 11 instead of Java 9).

Again, even if Easy Random fixes all deprecated reflection APIs usage (like `Field.isAccessible`,  `field.setAccessible`, etc) to remove this warning, you would still need to explicitly open your types for introspection if you use Java 9+.

**Comment by benas** [comment]
> Put only

> def jvmOpenModulesArgs = [
>         "--add-opens=java.base/java.util=ALL-UNNAMED",
> ]
> Don't solve problem.

The type you are trying to randomize `class SomeException extends RuntimeException` causes the field `java.lang.Throwable.backtrace` to be accessed through reflection, which is in `java.lang` and not `java.util`. So you need to open `java.lang` and not `java.util`:

```diff
-- --add-opens=java.base/java.util=ALL-UNNAMED
++ --add-opens=java.base/java.lang=ALL-UNNAMED
```

I tested this with JDK 11 and the warning disappeared. My [previous message](https://github.com/j-easy/easy-random/issues/385#issuecomment-597363143) seems to be inaccurate in that regard, sorry for that.

So to summarize: This warning does not appear because of an illegal access to JDK internal APIs in Easy Random's code, but because **you** are telling Easy Random to generate a random instance of a type that extends an internal class from the JDK. In this case, you need to open that type for introspection as mentioned in the [Java docs](https://docs.oracle.com/en/java/javase/11/tools/java.html#GUID-3B1CE181-CD30-4178-9602-230B800D4FAE):

```
If you must continue to use a component that requires illegal access,
then you can eliminate the warning messages by using one or more --add-opens
options to open only those internal packages to which access is required.
```

This is basically what I tried to explain in my previous messages.

> That said, we need to update Easy Random to replace the usage of Field.isAccessible (which is deprecated starting from Java 9) with field.canAccess(object).

This was done in 370d2f3bc03bcb5f9d9ddd030b3b05cd63c984e8.
